### ISSUE TITLE
NaN loss and/or adding training data causes dysync between iteration count in toml file and actual end of training.

### ISSUE THREAD

**Issue by pokepress** [neosr-project/neosr#88]
If one of the losses returns NaN, training aborts. If you then try to disable some losses and run again, it's possible that the training stops earlier than it should based on the iteration setting in the toml file. For example, I currently have a run that has an endpoint of 80000 iterations listed in the toml file, but it stops immediately after resuming from 60000 iterations believing that it's done. Looking at the logs:

```
Total epochs 2223 for total iters 80000.  
...
31-10-2024 04:50 PM | INFO: Resuming training from epoch: 2361, iter: 60000
31-10-2024 04:50 PM | INFO: AMP enabled.
31-10-2024 04:50 PM | INFO: Start training from epoch: 2361, iter: 60000
31-10-2024 04:50 PM | INFO: End of training. Time consumed: 0:00:00 
```
Basically, it thinks it's already completed enough epochs to finish the iterations required. I have been adding more files to the training data, but I think this is still possible without that. In either case, since the user can only really specify the number of iterations, I think it makes the most sense to use that as the barometer of when to stop.

**Comment by neosr-project** [comment]
Hi!

> If one of the losses returns NaN, training aborts

Yes, that is intentional. Since all losses are summed prior to backpropagation, NaN in a single loss will cause the optimizer to propagate NaN's. It shouldn't be happening though, please try the default configurations provided in the `options/` folder.

> but it stops immediately after resuming from 60000 iterations

Yes, that's a [known issue](https://github.com/neosr-project/neosr/wiki/FAQ,-Known-Issues-and-Roadmap#roadmap). Basically the `total_iter` option is based on epochs, not actual iters. So if you start training with a specific batch size and dataset size, and then resume with a lower iter per epoch, it can stop earlier than it should. However, it doesn't affect training, this option only stops training. The 'solution' is to simply increase the `total_iter` option and resume, for now.

> I think it makes the most sense to use that as the barometer of when to stop.

I agree, it should be fixed. Added it on my TODO list. Also, feel free to send PR's to the `dev` branch if you have python knowledge :+1:
